aten/src/ATen/cpu/vec256/vec256_complex_double_no_vec.h

```diff
@@ -0,0 +1,328 @@
+#pragma once
+
+#include <ATen/cpu/vec256/intrinsics.h>
+#include <ATen/cpu/vec256/vec256_base.h>
+#if defined(__AVX__) && !defined(_MSC_VER)
+#include <sleef.h>
+#endif
+
+namespace at {
+namespace vec256 {
+// See Note [Acceptable use of anonymous namespace in header]
+namespace {
+
+template <> class Vec256<std::complex<double>> {
+private:
+  std::complex<double> values[2] = {0};
+public:
+  using value_type = std::complex<double>;
+  static constexpr int size() {
+    return 2;
+  }
+  Vec256() {}
+  Vec256(std::complex<double> val) {
+    for (int i = 0; i != size(); i++) {
+      values[i] = val;
+    }
+  }
+  Vec256(std::complex<double> val1, std::complex<double> val2) {
+    values[0] = val1;
+    values[1] = val2;
+  }
+  template <int64_t mask>
+  static Vec256<std::complex<double>> blend(const Vec256<std::complex<double>>& a, const Vec256<std::complex<double>>& b) {
+    int64_t mask_ = mask;
+    Vec256 vec;
+    for (int i = 0; i < size(); i++) {
+      if (mask & 0x01) {
+        vec[i] = b[i];
+      } else {
+        vec[i] = a[i];
+      }
+      mask_ = mask_ >> 1;
+    }
+    return vec;
+  }
+  static Vec256<std::complex<double>> blendv(const Vec256<std::complex<double>>& a, const Vec256<std::complex<double>>& b,
+                               const Vec256<std::complex<double>>& mask) {
+    Vec256 vec;
+    int64_t buffer[size()];
+    mask.store(buffer);
+    for (int i = 0; i < size(); i++) {
+      if (buffer[i] & 0x01)
+       {
+        vec[i] = b[i];
+      } else {
+        vec[i] = a[i];
+      }
+    }
+    return vec;
+  }
+  static Vec256<std::complex<double>> arange(std::complex<double> base = 0., std::complex<double> step = 1.) {
+    return Vec256(base, base + step);
+  }
+  static Vec256<std::complex<double>> set(const Vec256<std::complex<double>>& a, const Vec256<std::complex<double>>& b,
+                            int64_t count = size()) {
+    Vec256 vec;
+    for (int i = 0; i < size(); i++) {
+      if (i < count) {
+        vec[i] = b[i];
+      } else {
+        vec[i] = a[i];
+      }
+    }
+    return vec;
+  }
+  static Vec256<std::complex<double>> loadu(const void* ptr, int64_t count = size()) {
+    Vec256 vec;
+    std::memcpy(vec.values, ptr, count * sizeof(std::complex<double>));
+    return vec;
+  }
+  void store(void* ptr, int count = size()) const {
+    std::memcpy(ptr, values, count * sizeof(std::complex<double>));
+  }
+  const std::complex<double>& operator[](int idx) const {
+    return values[idx];
+  }
+  std::complex<double>& operator[](int idx) {
+    return values[idx];
+  }
+  Vec256<std::complex<double>> map(std::complex<double> (*f)(const std::complex<double> &)) const {
+    Vec256<std::complex<double>> ret;
+    for (int i = 0; i != size(); i++) {
+      ret[i] = f(values[i]);
+    }
+    return ret;
+  }
+  Vec256<std::complex<double>> abs() const {
+    Vec256<std::complex<double>> ret;
+    for (int i = 0; i < size(); i++) {
+      ret[i] = std::abs(values[i]);
```

Comment:
It probably would have been ok to replace the `values[i] < 0 ? -values[i] : values[i]` in base with abs, you get the same code in each case (and then the complex overload works): https://godbolt.org/z/UjshUJ

Reply:
Thanks for introducing me to this website! The fact that you can show the assembly is the same for both approaches is amazingly helpful.